x-pack/legacy/plugins/infra/public/containers/logs/log_filter/log_filter_state.ts

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useState, useMemo } from 'react';
+import createContainer from 'constate';
+import { IIndexPattern } from 'src/plugins/data/public';
+import { esKuery } from '../../../../../../../../src/plugins/data/public';
+import { convertKueryToElasticSearchQuery } from '../../../utils/kuery';
+
+export interface KueryFilterQuery {
+  kind: 'kuery';
+  expression: string;
+}
+
+export interface SerializedFilterQuery {
+  query: KueryFilterQuery;
+  serializedQuery: string;
+}
+
+interface LogFilterInternalStateParams {
+  filterQuery: SerializedFilterQuery | null;
+  filterQueryDraft: KueryFilterQuery | null;
+}
+
+export const logFilterInitialState: LogFilterInternalStateParams = {
+  filterQuery: null,
+  filterQueryDraft: null,
+};
+
+export type LogFilterStateParams = Omit<LogFilterInternalStateParams, 'filterQuery'> & {
+  filterQuery: SerializedFilterQuery['serializedQuery'] | null;
+  filterQueryAsKuery: SerializedFilterQuery['query'] | null;
+  isFilterQueryDraftValid: boolean;
+};
+export interface LogFilterCallbacks {
+  setLogFilterQueryDraft: (expression: string) => void;
+  applyLogFilterQuery: (expression: string) => void;
+}
+
+export const useLogFilterState: (props: {
+  indexPattern: IIndexPattern;
+}) => [LogFilterStateParams, LogFilterCallbacks] = ({ indexPattern }) => {
+  const [state, setState] = useState(logFilterInitialState);
+  const { filterQuery, filterQueryDraft } = state;
+
+  const callbacks: LogFilterCallbacks = useMemo(() => {
+    const setLogFilterQueryDraft = (payload: KueryFilterQuery) =>
+      setState({ ...state, filterQueryDraft: payload });
```

Comment:
If the set operation depends on the `state` how about using the safer functional form to derive the new state

```suggestion
      setState(state => ({ ...state, filterQueryDraft: payload }));
```

That also means that the `useMemo` doesn't need to close over `state` from the outer scope an thus doesn't need the dependency.